### ISSUE TITLE
Add CORS headers to downloads enabling WebTorrent

### ISSUE THREAD

**Issue by wmhilton** [nodejs/build#678]
Howdy! Inspired by the recent downtime, I've been working on a way to make a resilient Node downloader using [WebTorrent](https://github.com/feross/webtorrent). Even if the primary download source becomes temporarily unavailable, the files would continue to be available on the bittorrent network. For my proof-of-concept [Download with WebTorrent Button](https://github.com/wmhilton/download-with-webtorrent-button/) to work with nodejs.org, I think all that is needed is to add Cross-Origin Resource Sharing (CORS) headers to the server config so it is possible to download files in `https://nodejs.org/dist/` via AJAX. I know that sounds weird, but how the torrent would work is if there are no other users seeding a torrent it would request the file from the source (via AJAX) to bootstrap the process. This also enables new torrents to be created automatically for files that don't have torrents the first time the button is clicked.

Obviously this isn't a panacea, because a lot of downloads are from 'nvm', WebTorrent only works on modern browsers that support WebRTC, and there are probably a host of other things to consider. But I love WebTorrent and love resilient distributed networking, and I thought adding CORS headers would be preferable to using a proxy service like [cors-anywhere](https://cors-anywhere.herokuapp.com/). You can see the demo of the button [here](https://wmhilton.com/download-with-webtorrent-button)

With that I'll open the floor to questions!

**Comment by feross** [comment]
I can't think of any security issues caused by adding CORS headers. It'll just make it possible for javascript on any website to download the installer files and use the data.

In addition to setting the `Access-Control-Allow-Origin: *` header on responses to GET request, it's important to also support the `OPTIONS` preflight method.

When a client sets a custom header like `Range` (to make http range requests) the browser first sends a "preflight" request to ask the server if the request should be allowed. (That's because certain methods like POST, or certain headers could be dangerous).

So, when a client sends an OPTIONS request, this would be a good response:

```
Access-Control-Allow-Methods: GET,HEAD
Access-Control-Allow-Headers: Range
Access-Control-Max-Age: 600
```

This allows GET and HEAD requests, allows the `Range` header, and caches this fact for 10 minutes in the client.

**Comment by wmhilton** [comment]
@feross That brings up an interesting point. Does the server have to support HTTP range requests to work as a webseed?

Because I noticed that some file extensions (tested in the https://nodejs.org/dist/v6.10.2/ directory) do *not* respect range requests:
- .txt, .asc, .sig, .lib

These file extensions *did* respect range requests:
- .gz, .xz, .7z, .zip, .msi, .exe

It also might be a file size rather than file extension thing. The text files and the .lib file that I tested were all smaller than 1.5MB.

**Comment by wmhilton** [comment]
> I can't think of any security issues caused by adding CORS headers.

That is correct. CORS is simply a way to bypass the same-origin policy, which prevents malicious pages from obtaining sensitive data on other sites by using AJAX and exploiting the fact the user is logged in with a cookie. Since the nodejs site, or at least the /dist folder, contains no sensitive data, there is absolutely no security advantage to not having CORS headers.

**Comment by jbergstroem** [comment]
@wmhilton: FWIW, we will shortly have cloudflare front our downloads; perhaps see how that's supported at their end as well.

**Comment by wmhilton** [comment]
I use the free level of Cloudflare for my sites, and while it adds a few headers I've never had a problem with it stripping or altering CORS headers. Cloudflare [passes Access-Control-Allow-Origin header through unaltered](https://support.cloudflare.com/hc/en-us/articles/200308847-Does-CloudFlare-support-Cross-origin-resource-sharing-CORS-).

**Comment by jbergstroem** [comment]
@wmhilton was rather talking about range requests; but yeah - the more testing the better!

**Comment by wmhilton** [comment]
Good thinking. I just tested, and I am able to do range requests on a file on my website that is cached by Cloudflare and hosted on github pages.

```sh
curl -X GET \
  http://wmhilton.com/hubo-js/data/hubo-urdf/meshes/Body_Torso.stl \
  -H 'cache-control: no-cache' \
  -H 'postman-token: 8c90a4b6-99cd-06bf-0f8c-6577f6e1bd99' \
  -H 'range: bytes=0-15'
```

**Comment by jbergstroem** [comment]
Ok, great that we have that confirmed. I have no issues with adding cors headers for downloads; just want to test out options configuration.

**Comment by wmhilton** [comment]
I still would love to see CORS headers added, but in the meantime (and because opening a Github issue with every website in the world isn't practical) I made a workaround service called [cors-buster](https://cors-buster.now.sh) that proxies HTTP range requests. I list both the canonical URL and the proxied one as webseeds in the output of [webtorrentify-server](https://webtorrentify.now.sh) so now my demo [download-with-webtorrent-button](https://wmhilton.com/download-with-webtorrent-button/) actually works!

The very first time it's used, this download button is much slower, because the file has to be downloaded once on https://webtorrentify.now.sh (if it isn't cached) to generate a torrent, and again through the https://cors-buster.now.sh proxy. However webtorrentify-server memoizes all its results, so the second time a link is clicked it should return the cached copy right away. When only one user is downloading node via WebTorrent, the only seed is the (cors-proxied) original URL so there is no bandwidth savings. But when you download on multiple browsers simultaneously, you start to see a huge speedup, because it starts downloading pieces from multiple neighbors in parallel. And most importantly, as the number of simultaneous downloads of the file increases, the bandwidth demand on the origin server actually *decreases*. In fact, the origin server could go down and nobody would notice.

**Comment by maclover7** [comment]
ping -- is this still something we want to do?

**Comment by wmhilton** [comment]
I'd still love it. :)

**Comment by feross** [comment]
Same – it's just a single header change and it would make it easier to download Node.js in new ways :)

**Comment by rvagg** [comment]
`nodejs.org` nginx config file is @ https://github.com/nodejs/build/tree/master/setup/www/resources/config if you want to PR to that, I can't think of any objections that might arise.

**Comment by wmhilton** [comment]
It looks like it already has a CORS header for `.json` files on [this line](https://github.com/nodejs/build/blame/master/setup/www/resources/config/nodejs.org#L59). Here's the current config:

```nginx
    location /dist {
        alias /home/dist/nodejs/release/;
        autoindex on;
        default_type text/plain;

        location ~ \.json$ {
            add_header access-control-allow-origin *;
        }
    }
```


@feross Can you help me figure out how we would change it? Here's my first draft:

```nginx
    location /dist {
        alias /home/dist/nodejs/release/;
        autoindex on;
        default_type text/plain;

        # Make it possible for javascript on any website to download the installer files and use the data.
        # (As of Nginx 1.7.5, add_header supports an "always" parameter which
        # allows CORS to work if the backend returns 4xx or 5xx status code.)
        add_header 'Access-Control-Allow-Origin' '*' always;
        add_header 'Access-Control-Allow-Methods' 'GET, HEAD, OPTIONS' always;
        add_header 'Access-Control-Allow-Headers' 'Range,Content-Type' always;
        if ($request_method = 'OPTIONS') {
            # Tell client that this pre-flight info is valid for 10 minutes
            add_header 'Access-Control-Max-Age' 600;
            add_header 'Content-Type' 'text/plain charset=UTF-8';
            add_header 'Content-Length' 0;
            return 204;
        }

        # JSON files are now covered by the general CORS settings directly above
        #location ~ \.json$ {
        #    add_header access-control-allow-origin *;
        #}
    }
```

**Comment by feross** [comment]
@wmhilton The proposed config looks good to me.

Just wondering – why is it necessary to allow the `Content-Type` header with `Access-Control-Allow-Headers`? It seems that since only GET, HEAD, and OPTIONS methods are allowed, the client will never need to set `Content-Type`, unless I'm missing something.

**Comment by wmhilton** [comment]
IIRC what happens is that some libraries automatically insert a `Content-Type` if you specify a particular body-parsing method (like `application/json` if you request JSON or `application/octet-stream` if you request binary) and that causes CORS errors. Because quote "The only allowed values for the Content-Type header are: application/x-www-form-urlencoded, multipart/form-data, and text/plain" ([source](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS))

You would think the client would just silently strip the Content-Type header from the CORS request, but IIRC it rejects the request before it gets a chance to leave the browser. :smh:

**Comment by wmhilton** [comment]
So its inclusion is to avoid getting "I use [insert name of AJAX library here] and it didn't work" issues.

**Comment by feross** [comment]
Yep, you're correct. In fact, this is what `simple-get` (used by WebTorrent) actually does.

So, this LGTM! Want to send the PR?

**Comment by wmhilton** [comment]
I guess I should!

**Comment by maclover7** [comment]
Closing in favor of the pull request, #1275
